### PR TITLE
Fix KMS encryption script to run in GCP cloud shell

### PR DESCRIPTION
**Pull request by Edwin-Pau**
Applied several fixes that enable the KMS encryption script under tools to be run from GCP cloud shell directly. A use-case would be if a user wanted to use the script to decrypt the terraform.tfvars file after completing the GCP quickstart deployment.

Signed-off-by: Edwin-Pau epau@teradici.com

### CODE REVIEW

**Comment by Edwin-Pau** [comment]
Please refer to the closed PR #112 for revision comments.

File: tools/kms_secrets_encryption.py [hunk 1]
```diff
@@ -7,13 +7,48 @@
...
+
+    if module_name is None:
+        module_name = pip_package
+
+    try:
+        module = importlib.import_module(module_name)
+        print(f"Successfully imported {module_name}.")
+
+    except ImportError:
+        install_cmd = f"pip3 install {pip_package} --user"
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: Can we add a prompt for user confirmation? Something like: "this script requires <pip-package-name> but is not available.  Install package by running "pip3 install {pip-package-name} --user"? (Y/N)" and if user enters N then quit.

**Review by shermanyin — CHANGES_REQUESTED** [review]

File: quickstart/gcp-cloudshell-quickstart.py [hunk 2]
```diff
@@ -101,6 +104,25 @@ def check_requirements():
         print('')
         sys.exit(1)
 
+    path = shutil.which('terraform')
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py:169]: where you check for terraform and the required python packages are in different functions, which seems a bit confusing.  I'd address the requirements in this function.  maybe rename the function to "ensure_requirements" instead of "check_requirements".  so 1) check for different requirements, if work is needed then ask for permission and set a flag to true.  we want to ask all the questions at the beginning of the script.  2) at the end of this function, go thru the flags and install the missing stuff.

File: quickstart/gcp-cloudshell-quickstart.py [hunk 3]
```diff
@@ -101,6 +104,25 @@ def check_requirements():
...
+            sys.exit(1)
+
+    try:
+        import googleapiclient.discovery
+        from google.cloud import kms_v1
+        from google.cloud.kms_v1 import enums
+        from google.api_core import exceptions as google_exc
+
+    except ImportError as err:
+        install_permission = input("Google Cloud module is not installed, proceed with installation (y/N)? ").strip().lower()
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py]: "Google Cloud module" sounds a bit weird.  I think those should be called "Python Google Client Libraries".  Same for other references to "modules"

File: quickstart/gcp-cloudshell-quickstart.py [hunk 4]
```diff
@@ -274,16 +296,29 @@ def tf_vars_create(ref_file_path, tfvar_file_path, settings):
 
 
 def terraform_install():
-    # Don't attempt to install unless needed, since it requires sudo
-    if not shutil.which('terraform'):
-        install_cmd = 'sudo python3 install-terraform.py'
-        subprocess.run(install_cmd.split(' '), check=True)
 
+    path = shutil.which('terraform')
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py]: no need for this check again if we move this code to under check_requirements() after the check for terraform

File: quickstart/gcp-cloudshell-quickstart.py (hunk 2 above)
Comment (shermanyin) [reply quickstart/gcp-cloudshell-quickstart.py:169]: Also need to check terraform version > 0.12.  If not then install a compatible version.

File: quickstart/gcp-cloudshell-quickstart.py [hunk 5]
```diff
@@ -101,6 +104,25 @@ def check_requirements():
         print('')
         sys.exit(1)
 
+    path = shutil.which('terraform')
+    if not path:
+        install_permission = input("Terraform is not installed, proceed with installation (y/N)? ").strip().lower()
+        if install_permission not in ('y', 'yes'):
+            print('Terraform is required for deployment. Exiting...')
+            sys.exit(1)
+
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py:189]: add a variable here like "install_terraform = true" and use that to determine if we need to perform installation at end of this function.

File: quickstart/gcp-cloudshell-quickstart.py [hunk 6]
```diff
@@ -101,6 +104,25 @@ def check_requirements():
...
+        from google.cloud import kms_v1
+        from google.cloud.kms_v1 import enums
+        from google.api_core import exceptions as google_exc
+
+    except ImportError as err:
+        install_permission = input("Google Cloud module is not installed, proceed with installation (y/N)? ").strip().lower()
+        if install_permission not in ('y', 'yes'):
+            print('Google Cloud modules are required for deployment. Exiting...')
+            sys.exit(1)
+
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py:166]: here add a flag that like install_google_libraries = true

Then go thru the flags and install missing stuff

File: quickstart/gcp-cloudshell-quickstart.py [hunk 7]
```diff
@@ -90,8 +90,10 @@
      web interface and manually created workstations. Resources not created by
      the Terraform scripts must be manually removed before Terraform can
      properly destroy resources it created.
-  2. In GCP cloudshell, change directory using the command "cd ~/cloud_deployment_scripts/{deployment_path}"
+  2. In GCP cloudshell, change directory using the command "cd ~/cloudshell_open/cloud_deployment_scripts/{deployment_path}"
   3. Remove resources deployed by Terraform using the command "terraform destroy". Enter "yes" when prompted.
+     Note: If Terraform was installed using this quickstart script, execute the following command instead.
+     "~/bin/terraform destroy"
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py]: not need for "if" here right? we know if we installed the script or not. Just display the proper command based on whether or not we installed terraform?

**Review by shermanyin — CHANGES_REQUESTED** [review]

**Comment by Edwin-Pau** [comment]
- Rebased on staging due to merge conflicts.
- Tested on Google Cloud Shell without Terraform installed.
- Tested on local machine without Google libraries installed.

File: quickstart/gcp-cloudshell-quickstart.py (hunk 5 above)
Comment (Edwin-Pau) [reply quickstart/gcp-cloudshell-quickstart.py:189]: I've made this flag a global one, since we want to reference it as part of the next_steps string.

File: quickstart/gcp-cloudshell-quickstart.py (hunk 6 above)
Comment (Edwin-Pau) [reply quickstart/gcp-cloudshell-quickstart.py:166]: For consistency sake, I've made this flag global as well.

File: quickstart/gcp-cloudshell-quickstart.py [hunk 8]
```diff
@@ -298,40 +359,20 @@ def tf_vars_create(ref_file_path, tfvar_file_path, settings):
...
-
 if __name__ == '__main__':
-    check_requirements()
+    ensure_requirements()
+
+    import googleapiclient.discovery
+    from google.cloud import kms_v1
+    from google.cloud.kms_v1 import enums
+    from google.api_core import exceptions as google_exc
+    print('Successfully imported Google libraries.\n')
```
Comment (Edwin-Pau) [inline quickstart/gcp-cloudshell-quickstart.py]: We need to import in main(), otherwise it can't find use the import references, unless we import and return modules in main, like in the encryption script. But that would be part of a future task which is refactoring this script to use the encryption script.
`Encrypting secrets...
Traceback (most recent call last):
  File "./gcp-cloudshell-quickstart.py", line 418, in <module>
    kms_client = kms_v1.KeyManagementServiceClient()
NameError: name 'kms_v1' is not defined`

File: tools/kms_secrets_encryption.py [hunk 9]
```diff
@@ -7,13 +7,61 @@
...
 
 SECRETS_START_FLAG = "# <-- Start of secrets section, do not edit this line. -->"
 
+
+def import_or_install_module(pip_package_name, module_name = None):
+    """A function that imports a Python top-level package or module. 
+    If the required package is not installed, it will install the package before importing it again.
+
+    Args:
+        pip_package_name (str): the name of the pip package to be installed
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: should be called pypi package.

File: quickstart/install-terraform.py [hunk 10]
```diff
@@ -3,18 +3,24 @@
...
 TEMP_DIR           = '/tmp'
 TERRAFORM_VERSION  = '0.12.3'
-TERRAFORM_BIN_DIR  = '/usr/local/bin'
+HOME               = os.path.expanduser('~')
+TERRAFORM_BIN_DIR  = f'{HOME}/bin'
 TERRAFORM_BIN_PATH = TERRAFORM_BIN_DIR + '/terraform'
 
+
 def terraform_install(version=TERRAFORM_VERSION):
+
```
Comment (shermanyin) [inline quickstart/install-terraform.py]: empty line not needed.

File: quickstart/install-terraform.py [hunk 11]
```diff
@@ -35,14 +42,19 @@ def terraform_install(version=TERRAFORM_VERSION):
...
 
 if __name__ == '__main__':
+    # Set up argparse
+    parser_description = ('Installs Terraform in the specified directory.')
 
+    parser = argparse.ArgumentParser(description=parser_description)
+    parser.add_argument('dir', help='specify the directory to install Terraform')
+    args = parser.parse_args()
+
+    TERRAFORM_BIN_DIR = args.dir
```
Comment (shermanyin) [inline quickstart/install-terraform.py]: so this overrides L18?

File: quickstart/gcp-cloudshell-quickstart.py [hunk 12]
```diff
@@ -40,11 +40,19 @@
 DEPLOYMENT_NAME = 'quickstart_deployment_' + iso_time
 CONNECTOR_NAME  = 'quickstart_cac_' + iso_time
 
+# Install flags
+INSTALL_TERRAFORM        = False
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py]: why do these have to be global if they are only used within one function?

File: quickstart/gcp-cloudshell-quickstart.py [hunk 13]
```diff
@@ -119,18 +217,18 @@ def quickstart_config_read(cfg_file):
...
-    1. Must not contain user's account name or display name
-    2. Must have 3 of the following categories:
-       - A-Z
-       - a-z
-       - 0-9
-       - special characters: (~!@#$%^&*_-+=`|\(){}[]:;"'<>,.?/)
-       - unicode characters
-
-    See: https://docs.microsoft.com/en-us/windows/security/threat-protection/security-policy-settings/password-must-meet-complexity-requirements
+Please enter a password for the Active Directory Administrator.
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py]: use this instead? https://docs.python.org/3/library/textwrap.html#textwrap.dedent

File: quickstart/gcp-cloudshell-quickstart.py [hunk 14]
```diff
@@ -88,20 +96,110 @@
...
+def ensure_google_libraries():
+    """A function that ensures Google libraries are installed. 
+
+    The function first tries to import the required Google libraries and if the required packages
+    are not installed, it will prompt the user to install these packages in the user's home directory. 
+    """
+
+    global INSTALL_GOOGLE_LIBRARIES
+
+    google_libraries_import = """
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py]: https://docs.python.org/3/library/textwrap.html#textwrap.dedent

File: quickstart/gcp-cloudshell-quickstart.py [hunk 15]
```diff
@@ -88,20 +96,110 @@
...
+    path = shutil.which('terraform')
+
+    if path:
+        cmd = 'terraform -v'
+        # Run the command 'terraform -v' and use the first line as the Terraform version
+        terraform_version = subprocess.run(cmd.split(' '),  stdout=subprocess.PIPE).stdout.decode('utf-8').splitlines()[0]
+        
+        print(f'Found {terraform_version} in {path}.\n')
+
+        if 'v0.12.' not in terraform_version:
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py]: You can check if v0.12. is in terraform_version, if so then just return the path.

Then you don't need the else and just proceed to install terrform without an extra if on L191.

Also returning the path will mean you avoid referencing the global variable.

**Review by shermanyin — CHANGES_REQUESTED** [review]

File: quickstart/install-terraform.py (hunk 11 above)
Comment (Edwin-Pau) [reply quickstart/install-terraform.py]: Yes, the command line argument would override the global bin dir. The script's default is the same as what we pass in at the moment, is this ok?

File: quickstart/gcp-cloudshell-quickstart.py (hunk 12 above)
Comment (Edwin-Pau) [reply quickstart/gcp-cloudshell-quickstart.py]: `INSTALL_TERRAFORM` is used towards the end to print the command string back to the user. I guess leaving the fully qualified path regardless of Terraform is installed or not would work too right? But I just thought it may be cleaner for the user if they can just use `terraform destroy` vs `/home/bin/terraform destroy`

Comment (Edwin-Pau) [reply quickstart/gcp-cloudshell-quickstart.py]: Global flags have been removed.

File: quickstart/gcp-cloudshell-quickstart.py (hunk 6 above)
Comment (Edwin-Pau) [reply quickstart/gcp-cloudshell-quickstart.py:166]: Flags have been removed since the logic to handle installing Terraform or Google libraries are now in their own functions.

File: tools/kms_secrets_encryption.py [hunk 16]
```diff
@@ -7,13 +7,61 @@
...
 SECRETS_START_FLAG = "# <-- Start of secrets section, do not edit this line. -->"
 
+
+def import_or_install_module(pypi_package_name, module_name = None):
+    """A function that imports a Python top-level package or module. 
+    If the required package is not installed, it will install the package before importing it again.
+
+    Args:
+        pypi_package_name (str): the name of the PyPI package to be installed
+        module_name (str): the import name of the module if it is different than the pip package name
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: not "pip package"

File: quickstart/install-terraform.py (hunk 11 above)
Comment (shermanyin) [reply quickstart/install-terraform.py]: Is there a scenario when the value assigned in L18 is used then?

Comment (Edwin-Pau) [reply quickstart/install-terraform.py]: Currently L18 would not be used - should it be initialized to None instead?

Or should I make the argparse directory argument an optional one? That way if no argument is specified to insta--terraform.py, then it would use the default directory in the script. Which way is better?

Comment (shermanyin) [reply quickstart/install-terraform.py]: How about remove the global, make it a required parameter for this file, and pass it into the terraform_install() function as a required parameter.

File: quickstart/gcp-cloudshell-quickstart.py [hunk 17]
```diff
@@ -124,15 +212,15 @@ def ad_password_get():
...
-       - A-Z
-       - a-z
-       - 0-9
-       - special characters: (~!@#$%^&*_-+=`|\(){}[]:;"'<>,.?/)
-       - unicode characters
+    - A-Z
+    - a-z
+    - 0-9
+    - special characters: (~!@#$%^&*_-+=`|\(){}[]:;"'<>,.?/)
+    - unicode characters
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py:219]: don't need to get rid of these spaces

File: quickstart/install-terraform.py [hunk 18]
```diff
@@ -3,30 +3,33 @@
...
 
+
 TEMP_DIR           = '/tmp'
 TERRAFORM_VERSION  = '0.12.3'
-TERRAFORM_BIN_DIR  = '/usr/local/bin'
-TERRAFORM_BIN_PATH = TERRAFORM_BIN_DIR + '/terraform'
 
-def terraform_install(version=TERRAFORM_VERSION):
+
+def terraform_install(terraform_bin_dir, version=TERRAFORM_VERSION):
```
Comment (shermanyin) [inline quickstart/install-terraform.py:19]: the function would make more sense if "version" is optional.  If it's not specified, install latest.  otherwise, install the specific version.  but let's leave this for a separate PR.

File: quickstart/gcp-cloudshell-quickstart.py [hunk 19]
```diff
@@ -88,20 +93,103 @@
...
+    path = shutil.which('terraform')
+
+    if path:
+        cmd = 'terraform -v'
+        # Run the command 'terraform -v' and use the first line as the Terraform version
+        terraform_version = subprocess.run(cmd.split(' '),  stdout=subprocess.PIPE).stdout.decode('utf-8').splitlines()[0]
+        
+        print(f'Found {terraform_version} in {path}.')
+
+        if 'v0.12.' in terraform_version:
```
Comment (shermanyin) [inline quickstart/gcp-cloudshell-quickstart.py:178]: this doesn't actually check for >=0.12.  For example, 0.13 would fail this check.

**Review by shermanyin — CHANGES_REQUESTED** [review]

**Comment by Edwin-Pau** [comment]
These changes have been pushed and merged from BitBucket, closing this PR.
